### PR TITLE
feat(js-sdk): Add API key

### PR DESCRIPTION
**Pull request by olivermrbl**
Closes CC-37

### CODE REVIEW

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 1]
```diff
@@ -21,12 +19,17 @@ export const useApiKey = (
   id: string,
   query?: Record<string, any>,
   options?: Omit<
-    UseQueryOptions<AdminApiKeyResponse, Error, AdminApiKeyResponse, QueryKey>,
+    UseQueryOptions<
+      HttpTypes.AdminApiKeyResponse,
+      Error,
+      HttpTypes.AdminApiKeyResponse,
```
Comment (riqwan) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]: q: shouldn't one of these be the allowed query params or `SelectParams`?

**Review by riqwan — APPROVED** [review]
one question, lgtm!

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 2]
```diff
@@ -21,12 +19,17 @@ export const useApiKey = (
   id: string,
   query?: Record<string, any>,
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]: Needs to be typed as well

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 3]
```diff
@@ -21,12 +19,17 @@ export const useApiKey = (
...
   options?: Omit<
-    UseQueryOptions<AdminApiKeyResponse, Error, AdminApiKeyResponse, QueryKey>,
+    UseQueryOptions<
+      HttpTypes.AdminApiKeyResponse,
+      Error,
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]:
```suggestion
      FetchError,
```

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 4]
```diff
@@ -38,16 +41,16 @@ export const useApiKeys = (
   query?: Record<string, any>,
   options?: Omit<
     UseQueryOptions<
-      AdminApiKeyListResponse,
+      HttpTypes.AdminApiKeyListResponse,
       Error,
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]: FetchError

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 5]
```diff
@@ -38,16 +41,16 @@ export const useApiKeys = (
   query?: Record<string, any>,
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]: Needs to be typed

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 6]
```diff
@@ -56,10 +59,14 @@ export const useApiKeys = (
 }
 
 export const useCreateApiKey = (
-  options?: UseMutationOptions<AdminApiKeyResponse, Error, CreateApiKeyReq>
+  options?: UseMutationOptions<
+    HttpTypes.AdminApiKeyResponse,
+    Error,
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]:
```suggestion
    FetchError,
```

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 7]
```diff
@@ -71,10 +78,14 @@ export const useCreateApiKey = (
 
 export const useUpdateApiKey = (
   id: string,
-  options?: UseMutationOptions<AdminApiKeyResponse, Error, UpdateApiKeyReq>
+  options?: UseMutationOptions<
+    HttpTypes.AdminApiKeyResponse,
+    Error,
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]: FetchError

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 8]
```diff
@@ -87,10 +98,10 @@ export const useUpdateApiKey = (
 
 export const useRevokeApiKey = (
   id: string,
-  options?: UseMutationOptions<AdminApiKeyResponse, Error, void>
+  options?: UseMutationOptions<HttpTypes.AdminApiKeyResponse, Error, void>
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]:
```suggestion
  options?: UseMutationOptions<HttpTypes.AdminApiKeyResponse, FetchError, void>
```

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 9]
```diff
@@ -102,10 +113,10 @@ export const useRevokeApiKey = (
 
 export const useDeleteApiKey = (
   id: string,
-  options?: MutationOptions<ApiKeyDeleteRes, Error, void>
+  options?: MutationOptions<HttpTypes.DeleteResponse<"api_key">, Error, void>
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]:
```suggestion
  options?: MutationOptions<HttpTypes.DeleteResponse<"api_key">, FetchError, void>
```

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 10]
```diff
@@ -118,14 +129,13 @@ export const useDeleteApiKey = (
 export const useBatchRemoveSalesChannelsFromApiKey = (
   id: string,
   options?: UseMutationOptions<
-    AdminApiKeyResponse,
+    HttpTypes.AdminApiKeyResponse,
     Error,
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]: FetchError

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 11]
```diff
@@ -118,14 +129,13 @@ export const useDeleteApiKey = (
...
     { sales_channel_ids: string[] }
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]: Should we add a type for this in the types package?

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 12]
```diff
@@ -142,13 +152,13 @@ export const useBatchRemoveSalesChannelsFromApiKey = (
 export const useBatchAddSalesChannelsToApiKey = (
   id: string,
   options?: UseMutationOptions<
-    AdminApiKeyResponse,
+    HttpTypes.AdminApiKeyResponse,
     Error,
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]:
```suggestion
    FetchError,
```

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx [hunk 13]
```diff
@@ -142,13 +152,13 @@ export const useBatchRemoveSalesChannelsFromApiKey = (
...
     { sales_channel_ids: string[] }
```
Comment (kasperkristensen) [inline packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]: Could add a type here again

File: packages/core/js-sdk/src/admin/api-key.ts [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+import { Client } from "../client"
+import { ClientHeaders } from "../types"
+
+export class ApiKey {
+  private client: Client
+  constructor(client: Client) {
+    this.client = client
+  }
+
+  async list(queryParams?: FindParams, headers?: ClientHeaders) {
```
Comment (kasperkristensen) [inline packages/core/js-sdk/src/admin/api-key.ts]: Should we not have a separate type for the query params here? I would expect you can do more filters then just FindParams

File: packages/core/js-sdk/src/admin/api-key.ts [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+      {
+        method: "DELETE",
+        headers,
+      }
+    )
+  }
+
+  async removeSalesChannels(
+    id: string,
+    body: { sales_channel_ids: string[] },
```
Comment (kasperkristensen) [inline packages/core/js-sdk/src/admin/api-key.ts]: Could use a type here

File: packages/core/js-sdk/src/admin/api-key.ts [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+        method: "DELETE",
+        headers,
+        body,
+      }
+    )
+  }
+
+  async addSalesChannels(
+    id: string,
+    body: { sales_channel_ids: string[] },
```
Comment (kasperkristensen) [inline packages/core/js-sdk/src/admin/api-key.ts]: Could use a type here

File: packages/core/js-sdk/src/admin/api-key.ts [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+  async list(queryParams?: FindParams, headers?: ClientHeaders) {
+    return await this.client.fetch<
+      PaginatedResponse<HttpTypes.AdminApiKeyListResponse>
+    >(`/admin/api-keys`, {
+      query: queryParams,
+      headers,
+    })
+  }
+
+  async create(body: HttpTypes.AdminCreateApiKey, headers?: ClientHeaders) {
```
Comment (kasperkristensen) [inline packages/core/js-sdk/src/admin/api-key.ts]: For other create/update endpoints we also allow passing params, to filter whats returned

**Comment by olivermrbl** [comment]
@kasperkristensen, thanks for review. Will address most of these. 

I tried to keep it simple (+ I used Cursor to do most of the work, which worked surprisingly well 😄 ) to get things moving on this project.

File: packages/admin-next/dashboard/src/hooks/api/api-keys.tsx (hunk 1 above)
Comment (olivermrbl) [reply packages/admin-next/dashboard/src/hooks/api/api-keys.tsx]: Yeah, I think you are right. It looks like we haven't used this much across the hooks, so there might be some clean-up work to do. I'll change this one 👍

**Pull request merged**